runtime/drivers/bigquery/sql_store.go

```diff
@@ -0,0 +1,220 @@
+}
+
+var _ drivers.RowIterator = &rowIterator{}
+
+func (r *rowIterator) Schema(ctx context.Context) (*runtimev1.StructType, error) {
+	if r.schema != nil {
+		return r.schema, nil
+	}
+
+	// schema is only available after first next call
+	r.next, r.nexterr = r.Next(ctx)
+	if r.nexterr != nil {
+		return nil, r.nexterr
+	}
+
+	fields := make([]*runtimev1.StructType_Field, len(r.bqIter.Schema))
+	for i, s := range r.bqIter.Schema {
+		dbt, err := toPB(s)
+		if err != nil {
+			return nil, err
+		}
+
+		fields[i] = &runtimev1.StructType_Field{Name: s.Name, Type: dbt}
+	}
+	r.schema = &runtimev1.StructType{Fields: fields}
+	return r.schema, nil
+}
+
+func (r *rowIterator) Next(ctx context.Context) ([]any, error) {
+	if r.next != nil || r.nexterr != nil {
+		next, err := r.next, r.nexterr
+		r.next = nil
+		r.nexterr = nil
+		return next, err
+	}
+
+	var row row = make([]any, 0)
+	if err := r.bqIter.Next(&row); err != nil {
+		if errors.Is(err, iterator.Done) {
+			return nil, drivers.ErrIteratorDone
+		}
+		return nil, err
+	}
+
+	return row, nil
+}
+
+func (r *rowIterator) Close() error {
+	return r.client.Close()
+}
+
+func (r *rowIterator) Size(unit drivers.ProgressUnit) (uint64, bool) {
+	if unit == drivers.ProgressUnitRecord {
+		return r.bqIter.TotalRows, true
+	}
+
+	return 0, false
+}
+
+type row []any
+
+var _ bigquery.ValueLoader = &row{}
+
+func (r *row) Load(v []bigquery.Value, s bigquery.Schema) error {
+	m := make([]any, len(v))
+	for i := 0; i < len(v); i++ {
+		if s[i].Type == bigquery.RecordFieldType || s[i].Repeated {
+			return fmt.Errorf("repeated or nested data is not supported")
+		}
+
+		m[i] = convert(v[i])
+	}
+	*r = m
+	return nil
+}
+
+// type conversion table for time
+// bigquery(format) -- duckdb -- internal pb type
+// TIMESTAMP -- TIMESTAMPTZ/TIMESTAMP WITH TIME ZONE -- Type_CODE_TIMESTAMP
+// DATETIME(9999-12-31 23:59:59.999999) -- TIMESTAMP/DATETIME -- Type_CODE_DATETIME
+// TIME([H]H:[M]M:[S]S[.DDDDDD|.F]) -- TIME -- Type_CODE_TIME
+// DATE(YYYY-[M]M-[D]D) -- DATE -- Type_CODE_DATE
+func toPB(field *bigquery.FieldSchema) (*runtimev1.Type, error) {
+	t := &runtimev1.Type{Nullable: !field.Required}
+	switch field.Type {
+	case bigquery.StringFieldType:
+		t.Code = runtimev1.Type_CODE_STRING
+	case bigquery.JSONFieldType:
+		t.Code = runtimev1.Type_CODE_JSON
+	case bigquery.IntervalFieldType:
+		t.Code = runtimev1.Type_CODE_STRING
+	case bigquery.GeographyFieldType:
+		t.Code = runtimev1.Type_CODE_STRING
+	case bigquery.FloatFieldType:
+		t.Code = runtimev1.Type_CODE_FLOAT64
+	case bigquery.NumericFieldType:
+		t.Code = runtimev1.Type_CODE_DECIMAL
+	// big numeric can have width upto 76 digits which can't be converted to DECIMAL type in duckdb
+	// which supports width upto 38 digits only.
+	case bigquery.BigNumericFieldType:
+		t.Code = runtimev1.Type_CODE_STRING
+	case bigquery.TimestampFieldType:
+		t.Code = runtimev1.Type_CODE_TIMESTAMP
+	case bigquery.DateTimeFieldType:
+		t.Code = runtimev1.Type_CODE_DATETIME
```

Comment:
I'm still unsure what value this adds – the BigQuery datetime value is read into a `time.Time` value (so still gets UTC time zone applied to it),  and `TIMESTAMP` and `DATETIME` are aliases in DuckDB. So it seems to me that using `CODE_TIMESTAMP` would exhibit equivalent behavior.

IMO the goal of the generic type representation isn't to represent all type names from all drivers 1:1, but more to reflect types with genuinely different behavior.

Reply:
So `TIMESTAMP`/`TIMESTAMPTZ` in duckdb and `DATETIME`/`TIMESTAMP ` in bigquery have difference that first type does not have any timezone associated with it so a value `1992-09-20 11:30:00` and `1992-09-20 11:30:00+00:00` even though may represent the same time internally in duckdb or other systems but is still different from user's pov. Quoting from BigQuery `TIMEZONE` is an absolute point in time but `DATETIME` is not. 
This is the information that we loose if use same internal proto type and than map it to same duckdb type. 

Even duckdb also removes the timezone information when presenting both values in CLI. 
![image](https://github.com/rilldata/rill/assets/12948312/2c00e5c8-1df8-4f7a-b5c6-f38d4409bf05)

If we don't store/understand this difference in types we can never show it accordingly in the application.